File: Shannon_entropy.py
import sys, math
from itertools import groupby, islice
from collections import Counter

def fasta_iter(fasta_name):
    """
    given a fasta file. yield tuples of header, sequence
    """
    fh = open(fasta_name)
    # ditch the boolean (x[0]) and just keep the header or sequence since
    # we know they alternate.
    faiter = (x[1] for x in groupby(fh, lambda line: line[0] == ">"))
    for header in faiter:
        # drop the ">"
        header = header.next()[1:].strip()
        # join all sequence lines to one.
        seq = "".join(s.strip() for s in faiter.next())
        yield header, seq

def sliding_window(seq, n=2):
    """Returns a sliding window (of width n) over data from the iterable
       s -> (s0,s1,...s[n-1]), (s1,s2,...,sn), ...                   
       this function courtesy http://stackoverflow.com/questions/6822725/rolling-or-sliding-window-iterator-in-python
    """
    it = iter(seq)
    result = tuple(islice(it, n))
    if len(result) == n:
        yield result    
    for elem in it:
        result = result[1:] + (elem,)
        yield result

def slidingWindow(sequence,winSize,step=1):
    """Returns a generator that will iterate through
    the defined chunks of input sequence.  Input sequence
    must be iterable."""
    
    # Verify the inputs
    try: it = iter(sequence)
    except TypeError:
        raise Exception("**ERROR** sequence must be iterable.")
    if not ((type(winSize) == type(0)) and (type(step) == type(0))):
        raise Exception("**ERROR** type(winSize) and type(step) must be int.")
    if step > winSize:
        raise Exception("**ERROR** step must not be larger than winSize.")
    if winSize > len(sequence):
        raise Exception("**ERROR** winSize must not be larger than sequence length.")
    
    # Pre-compute number of chunks to emit
    numOfChunks = ((len(sequence)-winSize)/step)+1
    
    # Do the work
    for i in range(0,numOfChunks*step,step):
        yield sequence[i:i+winSize]

def Shannon_entropy(seq, k):
    c = Counter()    
    #for kmer in sliding_window(seq, k):
    for kmer in slidingWindow(seq, k):
        new_count = c[kmer] + 1
        c[kmer] = new_count
    s = 0
    for i in c:
        #print i, c[i]
        p = float(c[i]) / (len(seq) - k + 1)
        s += p * math.log(p, 2)
    return (-s)

if __name__ == '__main__':
    fasta, k = sys.argv[1], sys.argv[2]

    for v in fasta_iter(fasta):
        print Shannon_entropy(v[1], int(k))

File: Shannon_entropy.go
package main

import (
	"fmt"
	"os"
	"bufio"
	"bytes"
	"io/ioutil"
	"math"
	"strconv"
)

func Shannon_entropy(seq []byte, winSize int, step int) float64 {
	m := make(map[string]int)
	str := string(seq)
	num := ((len(seq)-winSize)/step)+1
	for i := 0; i<num*step; i+=step {
		_, ok := m[str[i:i+winSize]]
		if !ok {
			m[str[i:i+winSize]] = 1
		} else {
			m[str[i:i+winSize]] += 1
		}
	}
	
	s := 0.0
	for _, value := range m {
		p := (float64(value) / float64(len(seq) - winSize + 1))
		s += p*math.Log2(p)
	}
	return (-s)
}

func main() {
	if len(os.Args) != 3 {
      panic("must provide sequence file.")
   	}
   
   	seq := ReadSequence(os.Args[1])
   	k, _ := strconv.Atoi(os.Args[2])
   	fmt.Println(Shannon_entropy(seq, k, 1))
}

func ReadSequence(file string) []byte {
   f, err := os.Open(file)
   if err != nil {
      panic(err)
   }
   defer f.Close()
   byte_array := make([]byte, 0)
   Ns := []byte("N")
   None := []byte("")
   if file[len(file)-6:] == ".fasta" || file[len(file)-3:] == ".fa" {
      scanner := bufio.NewScanner(f)
      for scanner.Scan() {
         line := scanner.Bytes()
         if len(line)>0 && line[0] != '>' {
            byte_array = append(byte_array, bytes.Replace(bytes.Trim(line,"\n\r "), Ns, None, -1)...)
         }
      }
   } else {
      byte_array, err = ioutil.ReadFile(file)
      if err != nil {
         panic(err)
      }
   }
   return byte_array
}